pytket/extensions/cutensornet/structured_state/classical.py

```diff
@@ -0,0 +1,136 @@
+        val = from_little_endian([bits_dict[b] for b in input_bits])  # type: ignore
+        # Check that the value is in the range
+        bits_dict[res_bit] = val >= op.lower and val <= op.upper
+
+    elif isinstance(op, ClassicalExpBox):
+        the_exp = op.get_exp()
+        result = evaluate_logic_exp(the_exp, bits_dict)
+
+        # The result is an int in little-endian encoding. We update the
+        # output register accordingly.
+        for b in bits:
+            bits_dict[b] = (result % 2) == 1
+            result = result >> 1
+        assert result == 0  # All bits consumed
+
+    elif op.type == OpType.Barrier:
+        pass
+
+    else:
+        raise NotImplementedError(f"Commands of type {op.type} are not supported.")
+
+
+def evaluate_logic_exp(exp: ExtendedLogicExp, bits_dict: dict[Bit, bool]) -> int:
+    """Recursive evaluation of a LogicExp."""
+
+    if isinstance(exp, int):
+        return exp
+    elif isinstance(exp, Bit):
+        return 1 if bits_dict[exp] else 0
+    elif isinstance(exp, BitRegister):
+        return from_little_endian([bits_dict[b] for b in exp])
+    else:
+
+        arg_values = [evaluate_logic_exp(arg, bits_dict) for arg in exp.args]
+
+        if exp.op in [BitWiseOp.AND, RegWiseOp.AND]:
+            return arg_values[0] & arg_values[1]
+        elif exp.op in [BitWiseOp.OR, RegWiseOp.OR]:
+            return arg_values[0] | arg_values[1]
+        elif exp.op in [BitWiseOp.XOR, RegWiseOp.XOR]:
+            return arg_values[0] ^ arg_values[1]
+        elif exp.op in [BitWiseOp.EQ, RegWiseOp.EQ]:
+            return int(arg_values[0] == arg_values[1])
+        elif exp.op in [BitWiseOp.NEQ, RegWiseOp.NEQ]:
+            return int(arg_values[0] != arg_values[1])
+        elif exp.op == BitWiseOp.NOT:
+            return 1 - arg_values[0]
+        elif exp.op == BitWiseOp.ZERO:
+            return 0
+        elif exp.op == BitWiseOp.ONE:
+            return 1
+        # elif exp.op == RegWiseOp.ADD:
+        #     return arg_values[0] + arg_values[1]
+        # elif exp.op == RegWiseOp.SUB:
+        #     return arg_values[0] - arg_values[1]
+        # elif exp.op == RegWiseOp.MUL:
+        #     return arg_values[0] * arg_values[1]
+        # elif exp.op == RegWiseOp.POW:
+        #     return int(arg_values[0] ** arg_values[1])
+        # elif exp.op == RegWiseOp.LSH:
+        #     return arg_values[0] << arg_values[1]
+        elif exp.op == RegWiseOp.RSH:
+            return arg_values[0] >> arg_values[1]
+        # elif exp.op == RegWiseOp.NEG:
+        #     return -arg_values[0]
+        else:
+            # TODO: Currently not supporting RegWiseOp's DIV, EQ, NEQ, LT, GT, LEQ,
+            # GEQ and NOT, since these do not return int, so I am unsure what the
+            # semantic is meant to be.
+            # TODO: Similarly, it is not clear what to do with overflow of ADD, etc.
+            # so I have decided to not support them for now.
```

Comment:
The semantics are not agreed upon across devices and emulators. For the time being, I'd rather error out as currently done. The current implementation should be capable of running any circuit that comes from a QASM file (parsed via pytket.qasm) using the standard QASM library. These extra operators on registers come from our internal extension to QASM, `hqslib1.inc`.